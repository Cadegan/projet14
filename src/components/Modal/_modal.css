.modal{position:fixed;z-index:99;left:0;top:0;width:100%;height:100%;overflow:auto;background-color:rgba(0,0,0,.3);-webkit-backdrop-filter:blur(2px);backdrop-filter:blur(2px)}.modalWrapper{padding:16px 20px;width:300px;margin:auto;background-color:#fff;border-radius:5px;position:fixed;top:50%;left:50%;transform:translate(-50%, -50%);box-shadow:0px 0px 10px 0px rgba(0,0,0,.3)}.closeModalBnt{cursor:pointer}