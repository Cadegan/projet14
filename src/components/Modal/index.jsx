import { useEffect } from "react";
import PropTypes from "prop-types";
import cross from "./assets/cross.svg";

/**
 * Modal component that displays a message in a modal window.
 *
 * @param {Object} props - The component props.
 * @param {boolean} props.openModal - Flag indicating whether the modal is open or not.
 * @param {Function} props.closeModal - Function to close the modal.
 * @param {string} props.message - The message to display in the modal.
 */
const Modal = ({ openModal, closeModal, message }) => {
  /**
   * This is a hook that is used to detect if the modal is open or not. If it is open, it will add a
   * style to the body of the document to hide the overflow. If it is not open, it will remove the
   * style from the body of the document.
   */
  useEffect(() => {
    document.body.style.overflow = openModal ? "hidden" : "";
  }, [openModal]);

  if (!openModal) return null;

  return (
    <div onClick={closeModal} data-testid={"closeModal"} className="modal">
      <div
        className="modalWrapper"
        data-testid={"modalWrapper"}
        onClick={(e) => {
          e.stopPropagation();
        }}
      >
        <div className="modalContent">
          <span data-testid={"message"}>{message}</span>
        </div>
        <button
          className="closeModalBnt"
          onClick={closeModal}
          data-testid={"closeModalBnt"}
        >
          <img src={cross} className="crossImage" alt="cross"></img>
        </button>
      </div>
    </div>
  );
};

/**
 * Proptypes for the Modal component.
 */
Modal.propTypes = {
  openModal: PropTypes.bool,
  closeModal: PropTypes.func,
  message: PropTypes.string,
};

export default Modal;
