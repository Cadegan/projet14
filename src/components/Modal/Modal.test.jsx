import { render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import React from "react";
import Modal from "./index";

describe("Modal", () => {
  test("calls useEffect with the correct arguments when openModal changes", () => {
    /* Spying on the useEffect hook. */
    const spy = jest.spyOn(React, "useEffect");
    /* Destructuring the `rerender` function from the `render` function. */
    const { rerender } = render(<Modal openModal={false} />);
    /* Re-rendering the component with the new props. */
    rerender(<Modal openModal={true} />);
    /* Checking that the useEffect hook was called with the correct arguments. */
    expect(spy).toHaveBeenCalledWith(expect.any(Function), [true]);
    /* It restores the original function. */
    spy.mockRestore();
  });

  test("returns null if openModal is false", () => {
    const { container } = render(<Modal openModal={false} />);
    expect(container).toBeEmptyDOMElement();
  });

  test("renders the correct HTML content when openModal is true", () => {
    render(<Modal openModal={true} />);
    expect(screen.getByTestId("closeModal")).toBeInTheDocument();
    expect(screen.getByTestId("modalWrapper")).toBeInTheDocument();
    expect(screen.getByTestId("closeModalBnt")).toBeInTheDocument();
  });

  test("calls closeModal when the user clicks the Close button or the modal background area", () => {
    const closeModal = jest.fn();
    render(<Modal openModal={true} closeModal={closeModal} />);
    fireEvent.click(screen.getByTestId("closeModalBnt"));
    expect(closeModal).toHaveBeenCalled();
    fireEvent.click(screen.getByTestId("closeModal"));
    expect(closeModal).toHaveBeenCalled();
  });

  test("correctly displays the message", () => {
    const message = "This is a test message";
    render(<Modal openModal={true} children={message} />);
    expect(screen.getByTestId("message")).toHaveTextContent(message);
  });
});
