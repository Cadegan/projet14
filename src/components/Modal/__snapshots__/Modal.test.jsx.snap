// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal has the correct PropTypes 1`] = `
<div>
  <div
    class="modal"
  >
    <div
      class="modalWrapper"
    >
      <div
        class="modalContent"
      >
        <span>
          This is a test message
        </span>
      </div>
      <button
        class="closeModalBnt"
      >
        <img
          alt="cross"
          class="crossImage"
          src="cross.svg"
        />
      </button>
    </div>
  </div>
</div>
`;
